src/PrettyPrinting.jl

```diff
@@ -1591,23 +1591,81 @@ function _write_line(io::IOCustom, str::AbstractString)
     spaceleft = c - ind - io.printed
   end
   #@show spaceleft
-  firstlen = min(spaceleft, length(str))
-  firststr = str[1:firstlen]
+  # The following code deals with line wrapping of Unicode text, including
+  # double-width symbols and more.
+  _graphemes = Base.Unicode.graphemes(str)
+  firstlen = min(spaceleft, length(_graphemes))
+  # make an iterator over valid indices
   if io.lowercasefirst
-    written += write(io.io, lowercasefirst(firststr))
+    firstiter = Base.Iterators.take(_graphemes, firstlen)
+    firstiter = join(firstiter)
+    firstiter = Base.Unicode.graphemes(lowercasefirst(firstiter))
     io.lowercasefirst = false
   else
-    written += write(io.io, firststr)
+    firstiter = Base.Iterators.take(_graphemes, firstlen)
     io.lowercasefirst = false
   end
-  io.printed += textwidth(firststr)
-  reststr = str[firstlen + 1:end]
-  it = Iterators.partition(1:textwidth(reststr), c - ind > 0 ? c - ind : c)
-  for i in it
+  restiter = Base.Iterators.drop(_graphemes, firstlen)
+  firststr = join(firstiter)
+  if length(firstiter) == textwidth(firststr)
+    written += write(io.io, firststr)
+    io.printed += textwidth(firststr)
+  else
+    #firstline is wider than number of graphemes
+    partcollect = collect(firstiter)
+    printstr = ""
+    j = 1
+    while textwidth(printstr) < (c - ind > 0 ? c - ind : c)
+      printstr *= partcollect[j]
```

Comment:
Note that this is essentially a quadratic algorithm now that could be linear (luckily bounded by the terminal width), because `textwidth` is called on longer and longer strings. 
```suggestion
    j = 1
    width = 0
    while width < (c - ind > 0 ? c - ind : c)
      printstr *= partcollect[j]
      width += textwidth(partcollect[j])
```